text/0000-min-target-api-version.md

```diff
@@ -0,0 +1,205 @@
+
+* set your crate's minimum API version higher than that of the standard library. 
+* change certain implementation details of your crate depending on what a downstream user sets their minimum API version to be.  
+* have some sensible compiler error if users of your crate require a lower minimum API version that you require.
+
+This is where certain `cfg` predicates come in that allow you to specify what the minimum target APIs are. These include the following:
+* `min_windows_build_version`: the [minimum build version](https://gaijin.at/en/infos/windows-version-numbers) of Windows.
+* `min_libc_version`: the version of libc in use.
+* `min_kernel_version`: the version of the kernel in use.
+* `min_macos_version`: The minium version of macOS in use.
+
+
+Let's use `min_windows_build_number` as an example. It should be clear how this example would be extended to the other `cfg` predicates. The `min_windows_build_number` allows you to conditionally compile code based on the set minimum API version. For example an implementation of mutex locking on Windows might look like this:
+
+```rust
+pub unsafe fn unlock(&self) {
+    *self.held.get() = false;
+    if cfg!(min_windows_build_version >= "6.0.6000") { // API version greater than Vista
+        c::ReleaseSRWLockExclusive(raw(self)) // Use the optimized ReleaseSRWLockExclusive routine
+    } else {
+        (*self.remutex()).unlock()  // Fall back to an alternative that works on older Windows versions
+    }
+}
+```
+
+End users can set the `min_windows_build_version` in the Cargo configuration file `.cargo/config` under the [`target key`](https://doc.rust-lang.org/cargo/reference/config.html#target) like so:
+
+```toml
+[target.x86_64-pc-windows-msvc]
+min_windows_build_version = "6.0.6000" # Vista 
+```
+
+When using `rustc` directly, this option can be provided as a command line option: `rustc --cfg 'min_windows_build_version="6.0.6000"'`.
+
+If an end user sets their `min_windows_build_version` to an incompatible version then the user receives an error. For instance, in the example above where the user is setting their `min_windows_build_version` to Windows Vista, they will receive an error when linking with the standard library which imposes Windows 7 as its `min_windows_build_version` by default for the `x86_64-pc-windows-msvc` target.
+
+If a library does not explicitly set its `min_windows_build_version`, it will automatically be set to the largest `min_windows_build_version` of all of its transitive dependencies.
+
+When user's code checks against a `min_windows_build_version`, a `>=` check is performed. This means that if the `min_windows_build_version` is "10.0.10240" and the user specifies `#[cfg(min_windows_build_version >= "6.2.9200"]` then the `cfg` predicate will be true and the code it marks will be compiled, since "10.0.10240" is a greater version than "6.2.9200".
```

Comment:
The `min_` prefix in the proposed cfg labels and the `>=` operator seem duplicate each other in the meaning. Since `>=` is a new syntax (as is acknowledged later) and would most likely need its own rfc, I think we want to avoid `>=` in this rfc.

(besides, if we push for `>=`, what would then the behaviour of `#[cfg(min_windows_build_version="6.2.9200")]` be?)

Reply:
The semantics of the `min_*_version` is that marked code should compile when the set minimum supported build version is at least equal to that version. i.e., if code is marked with `#[cfg(min_windows_build_version >= "windows7")]`, it should compile if the `min_windows_build_version` is set to "windows10". It has been argued that a plain "=" implies that the code would only compile when the `min_windows_build_version` matches the version provided exactly.

Reply:
The Zulip thread linked up top goes back and forth on `>=` a couple of times. It might be worth revisiting some of the arguments there to help move the discussion forward and save circling around the same points.